File: Tp2/BubbleSort/times.cpp
//Importando bibliotecas basicas
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

//Criando o tipo boolean
#define bool short
#define true 1
#define false 0

//Constantes globais
const long int TAM = 1000000;
const int TAMmenor = 1000;
#define tamzinho 300

//Strutura para armazenar os arquivos
typedef struct TIMES{
  char nomes[tamzinho];
  char liga[tamzinho];
}TIMES;

typedef struct Ordem{
  TIMES times;
}Ordem;


//Funcao para verificar se eh o fim
bool ehFim(char entrada[]){
  bool fim = true;

  if(entrada[0] != 'F' || entrada[1] != 'I' || entrada[2] != 'M'){
    fim = false;
  }
  return fim;
}


//Funcao para encontrar um texto em uma frase
bool find(char texto[], char procura[], int *resp){
  bool encontrar = false;
  long int tam = strlen(texto);
  int tamProcura = strlen(procura);
  int posProcura = -1;
  bool table = true;

  for(long int i = 0; i < tam; i++){
    if(texto[i] == procura[0] && table){
      encontrar = true;
      table = false;
      for(int j = 0; j < tamProcura; j++){
        if(texto[i+j] != procura[j]){
          encontrar = false;
          j = tamProcura;
        }
        else{
          table = true;
        }
      }
    }
    if(encontrar){
      posProcura = i + tamProcura;
      i = tam;
    }
  }
  *resp = posProcura;
  return encontrar;
}


//Funcao para retirar a tag table de todo o html
void limpandoEntrada(char entrada[], char textoLimpo[]){
  int tam = strlen(entrada);
  int posProcura = 0;
  int posFIM = 0;
  int pos = 0;

  find(entrada, "vcard", &posProcura);
  find(&entrada[posProcura], "</table>", &posFIM);
  for(int i = posProcura; i < posFIM+10000; i++){
    if(entrada[i] == '\n') {continue;}
    textoLimpo[pos] = entrada[i];
    pos++;
  }
}


//Removedor de tags HTML
char* removerTags(char entrada[]){
  bool chaves = true;
  char* string;
  int pos = 0;

  for (int i = 0; i < strlen(entrada); i++)
  {
    if (entrada[i] == '<')
      chaves = true;
    if (!chaves)
    {
      entrada[pos] = entrada[i];
      pos++;
    }
    if (entrada[i] == '>')
      chaves = false;
  }
  entrada[pos] = '\0';

  bool Ecomercial = false;
  int pos2 = 0;
  for (int i = 0; i < strlen(entrada); i++)
  {
    if (entrada[i] == '&')
    {
      Ecomercial = true;
    }
    if (!Ecomercial)
    {
      entrada[pos2] = entrada[i];
      pos2++;
    }
    if (entrada[i] == ';')
    {
      Ecomercial = false;
    }
  }
  entrada[pos2] = '\0';

  string = entrada;
  return string;
}

//Funcao para procurar os itens dos times
bool procurarItens(char entrada[], char procurarInicio[], char procurarFinal[], char resp[]){
  int posI;
  int posF;
  bool encontrar = false;
  encontrar = find(entrada, procurarInicio, &posI);
  if(encontrar){
    find(&entrada[posI], procurarFinal, &posF);
  }
  int j = 0;

  if(encontrar){
    for(int i = posI; i < posI+posF-strlen(procurarFinal); i++){
      resp[j] = entrada[i];
      j++;
    }
    resp[j] = '\0';
}
  strcpy(resp, removerTags(resp));

  return encontrar;
}


//Funcao para tirar pontuacao da capacidade
void consertarCapacidade(char capacidade[]){
  int pos = 0;

  for(int i = 0; i < strlen(capacidade); i++){
    if(capacidade[i] != ',' && capacidade[i] != '.'){
      capacidade[pos] = capacidade[i];
      pos++;
    }
  }
  capacidade[pos] = '\0';
}

//Funcao para printar as respostas na tela
void printar(char texto[]){

  for(int i = 0; i < strlen(texto); i++){
    printf("%c", texto[i]);
  }
  printf(" ## ");

}


//Funcao para ordenar a data
void ordenarData(char entrada[]){

  char ano[4], mes[2], dia[2];
  ano[0] = entrada[0];
  ano[1] = entrada[1];
  ano[2] = entrada[2];
  ano[3] = entrada[3];
  mes[0] = entrada[5];
  mes[1] = entrada[6];
  dia[0] = entrada[8];
  dia[1] = entrada[9];

  entrada[0] = dia[0];
  entrada[1] = dia[1];
  entrada[2] = '/';
  entrada[3] = mes[0];
  entrada[4] = mes[1];
  entrada[5] = '/';
  entrada[6] = ano[0];
  entrada[7] = ano[1];
  entrada[8] = ano[2];
  entrada[9] = ano[3];
  entrada[10] = '\0';

  //Consertar casos onde nao ha data
  for(int i = 0; i < strlen(entrada); i++){
    if(i == 2 || i == 5){
      //Nao fazer nada
    }
    else{
      if(entrada[i] < '0' || entrada[i] > '9'){
        entrada[i] = '0';
      }
    }
  }

  //Retirando numeros maiores que 12 no mes e maiores que 31 no dia
  int auxDia;
  int auxMes;

  auxDia = ((int)dia[0] - 48) + ((int)dia[1] - 48);
  auxMes = ((int)mes[0] - 48) + ((int)mes[1] - 48);

  if(auxDia > 31 || auxMes > 12){
    entrada[0] = '0';
    entrada[1] = '0';
    entrada[3] = '0';
    entrada[4] = '0';
  }
}


//Funcao para pegar as datas
void resolverDatas(char entrada[]){

  char ano[4], mes[2], dia[2];
  int pos = 0;
  for(int i = 0; i < strlen(entrada); i++){
    if(entrada[i] == '('){
      for(int j = 1; j < strlen(entrada); j++){
        entrada[pos] = entrada[i+j];
        pos++;
        if(entrada[i+j] == ')'){
        j = strlen(entrada);
        i = strlen(entrada);
        }
      }
    }
  }
  entrada[pos-1] = '\0';

  ordenarData(entrada);
}

void printarListaOrdenada(Ordem time[], int quantTimes){
  for (int i = 0; i < quantTimes; i++){
    printf("%s\n", time[i].times.nomes);
  }
}


void bolha(Ordem lista[], int* comparacoes, int* movimentacoes, int tam){
  Ordem aux;
  for(int i = 0; i < tam; i++){
    for(int j = i; j < tam -1; j++){
      comparacoes++;
      if(strcmp(lista[j].times.liga, lista[j+1].times.liga) > 0){
        aux = lista[j];
        lista[j] = lista[j+1];
        lista[j+1] = aux;
        movimentacoes++;
      }
    }
  }
}


/*

  FUNCOES PRINCIPAIS DO PROGRAMA

*/

//Funcao para organizar todo o codigo
//Ordenando as execucoes
void ORQUESTRADOR(char entrada[], Ordem lista[], int listaAtual, bool print){

  //Abrindo arquivo{
  FILE *arq;
  arq = fopen(entrada, "rb");
  char texto[TAM];

  //char texto[TAM];
  char textoLimpo[TAM];

  //Lendo o arquivo
  fread(texto, TAM, sizeof(char), arq);

  //Removendo itens inuteis do texto
  limpandoEntrada(texto, textoLimpo);


  //Procurar itens
  //Variaveis para armazenar as caracteristicas do time
  struct times{
    char nomeTime[TAMmenor];
    char apelidoTime[TAMmenor];
    char nomeEstadio[TAMmenor];
    char tecnico[TAMmenor];
    char liga[TAMmenor];
    char capacidade[TAMmenor];
    char data[TAMmenor];
    long int tamanhoArquivo;
  }time;

  /*
  //Zerar todas as variaveis
  memset(time.nomeTime, '\0', sizeof(time.nomeTime));
  memset(time.apelidoTime, '\0', sizeof(time.apelidoTime));
  memset(time.nomeEstadio, '\0', sizeof(time.nomeEstadio));
  memset(time.tecnico, '\0', sizeof(time.tecnico));
  memset(time.liga, '\0', sizeof(time.liga));
  memset(time.capacidade, '\0', sizeof(time.capacidade));
  memset(time.data, '\0', sizeof(time.data));
  */

  
  //Descobrir tamanho do Arquivo
  fseek(arq, 0, SEEK_END);
  time.tamanhoArquivo = ftell(arq);
  fseek(arq, 0, SEEK_SET);
  
  //Funcoes para procurar as caracteristicas dos times
  //printf("%s\n\n", textoLimpo);
  procurarItens(textoLimpo, "Full name", "</td></tr>", time.nomeTime);
  procurarItens(textoLimpo, "Nickname", "</td></tr>", time.apelidoTime);
  procurarItens(textoLimpo, "Ground", "</td></tr>", time.nomeEstadio);
  if (procurarItens(textoLimpo, "Manager", "</td>", time.tecnico)){}
  else if (procurarItens(textoLimpo, "Coach", "</td>", time.tecnico)) {}
  else {procurarItens(textoLimpo, "Head coach", "</td>", time.tecnico);}
  procurarItens(textoLimpo, "League", "</td></tr>", time.liga);
  procurarItens(textoLimpo, "Capacity", "</td></tr>", time.capacidade);
  procurarItens(textoLimpo, "Founded", "</td></tr>", time.data);
  //strcpy(time.tecnico, "0");	
 

  //Formatando prints
  consertarCapacidade(time.capacidade);
  resolverDatas(time.data);

  //Funcao para printar na tela os resultados
  if(print){
    printar(time.nomeTime);
    printar(time.apelidoTime);
    printar(time.data);
    printar(time.nomeEstadio);
    printar(time.capacidade);
    printar(time.tecnico);
    printar(time.liga);
    printar(entrada);
    printf(" %ld ## ", time.tamanhoArquivo);
    printf("\n");
  }

  /*
    Nome do time ## Apelido ## dia/mes/ano ## Estadio ## Capacidade ## Tecnico ## Liga ## Arquivo ## Bytes do arquivo ##\n
  */

 /*
	Sortings
 */
  int i;
  for(i = 0; i < strlen(time.tecnico); i++){
    lista[listaAtual].times.liga[i] = time.liga[i];
  }
	lista[listaAtual].times.liga[i] = '\0';
}

void SORT(char entrada[], bool ordenar, Ordem lista[], int listaAtual){
  //Executando os arquivos    
  if(!ordenar)
    ORQUESTRADOR(entrada, lista, listaAtual, false);

  /*-----
  ORDENAR APENAS QUANDO TODOS OS ARQUIVOS FOREM LIDOS
  -------*/

  /*if(ordenar){
    FILE *arq;
    arq = fopen("matrícula_bolha.txt", "wb");

    fprintf(arq, "%s", "650888\t");

    -----
    SORTING
    ------

    int comparacoes = 0;
    int movimentacoes = 0;

    ----------
    CRONOMETRANDO TEMPO DE EXECUCAO DA ORDENACAO
    ----------

    clock_t inicio = clock();

    bolha(lista, &movimentacoes, &comparacoes, listaAtual - 1);

    clock_t final = clock();
    double total = (double) (final - inicio) / CLOCKS_PER_SEC;

    //Numero de comparações (do array)
    fprintf(arq, "%d\t", comparacoes);

    //Numero de movimentações (do array)
    fprintf(arq, "%d\t", movimentacoes);

    //Tempo gasto de execucao
    fprintf(arq, "%lf\t", total);
  }*/
}

//O fgets adiciona um '\0' no final do input
void consertarFgets(char entrada[]){
  int tam = strlen(entrada);
  entrada[tam-1] = '\0';
}


int main(){
  char entrada[TAM];
  fgets(entrada, TAM, stdin);
  //Contador de numero arquivos
  Ordem ordem[TAMmenor];
  int numArquivo = 0;

  while(!ehFim(entrada)){
    consertarFgets(entrada);
    //Escrever nome do arquivo na struct

    //Salvar o nome dos arquivos em uma struct
    strcpy(ordem[numArquivo].times.nomes, entrada);

    SORT(entrada,  false, ordem, numArquivo);
    numArquivo++;
    fgets(entrada, TAM, stdin);
  }

  SORT(entrada, true, ordem, numArquivo);

  for (int i = 0; i < numArquivo; i++)
  {
    ORQUESTRADOR(ordem[i].times.nomes, ordem, i, true);
  }

  return 0;
}

File: Tp2/BubbleSort/Times.java
import java.util.Scanner;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.File;

class Times extends Time{
    public static void main(String[] args) throws IOException{
        Scanner leitor = new Scanner(System.in);
        String entrada = leitor.nextLine();
        Time[] time = new Time[50];
        int numArq = 0;

        while(!ehFim(entrada)){
            try{
                time[numArq] = new Time();
                time[numArq].ler(entrada);
                numArq++;
            }
            catch(IOException erro){
                throw new IOException("Não foi possivel abrir o arquivo!");
            }
            entrada = leitor.nextLine();
        }

        for(int i = 0; i < numArq; i++){
            System.out.println(time[i].nome);
        }
    }

    public static boolean ehFim (String entrada){
        boolean fim = true;

        if(entrada.charAt(0) != 'F' || entrada.charAt(1) != 'I' || entrada.charAt(2) != 'M')
            fim = false;

        return fim;
    }
}


class Time{
    protected String nome, apelido, estadio, tecnico, liga, nomeArquivo;
    protected int capacidade, fundacaoDia, fundacaoMes, fundacaoAno;
    protected long paginaTam;

    public Time() {
        nome = apelido = estadio = tecnico = liga = nomeArquivo = "";
        paginaTam = capacidade = fundacaoDia = fundacaoMes = fundacaoAno = 0;
    }

    public Time(String nome, String apelido, String estadio, String tecnico, String liga, String nomeArquivo, int capacidade, int fundacaoDia, int fundacaoMes, int fundacaoAno, long paginaTam){
        this.nome = nome;
        this.apelido = apelido;
        this.estadio = estadio;
        this.tecnico = tecnico;
        this.liga = liga;
        this.nomeArquivo = nomeArquivo;
        this.capacidade = capacidade;
        this.fundacaoAno = fundacaoAno;
        this.fundacaoMes = fundacaoMes;
        this.fundacaoDia = fundacaoDia;
        this.paginaTam = paginaTam;
    }

    public String toString() {
        String resp = nome + " ## " + apelido + " ## ";
        resp += ((fundacaoDia > 9) ? "" : "0") + fundacaoDia + ((fundacaoMes > 9) ? "/" : "/0") + fundacaoMes + "/"
                + fundacaoAno + " ## ";
        resp += estadio + " ## " + capacidade + " ## " + tecnico + " ## " + liga + " ## " + nomeArquivo + " ## "
                + paginaTam + " ## ";
        return resp;
    }

    public String removePunctuation(String campo) {
        campo = campo.replace(".", "");
        campo = campo.replace(",", "");
        campo = campo.replace(";", "");
        String resp = "";
        for (int i = 0; i < campo.length(); i++) {
            if (Character.isDigit(campo.charAt(i)))
                resp += campo.charAt(i);
            else
                i = campo.length();
        }
 
        return resp;
    }

    public int getMes(String campo) {
        int resp = 0;
        if (campo.contains("january") == true) {
            resp = 1;
        } else if (campo.contains("february") == true) {
            resp = 2;
        } else if (campo.contains("march") == true) {
            resp = 3;
        } else if (campo.contains("april") == true) {
            resp = 4;
        } else if (campo.contains("may") == true) {
            resp = 5;
        } else if (campo.contains("june") == true) {
            resp = 6;
        } else if (campo.contains("july") == true) {
            resp = 7;
        } else if (campo.contains("august") == true) {
            resp = 8;
        } else if (campo.contains("september") == true) {
            resp = 9;
        } else if (campo.contains("october") == true) {
            resp = 10;
        } else if (campo.contains("november") == true) {
            resp = 11;
        } else if (campo.contains("december") == true) {
            resp = 12;
        }
        return resp;
    }

    public String getSubstringEntre(String s, String antes, String depois) {
        String resp = "";
        int posInicio, posFim;
 
        posInicio = s.indexOf(antes) + antes.length();
 
        if (antes.compareTo(depois) != 0) {
            posFim = s.indexOf(depois);
        } else {
            posFim = s.indexOf(depois, posInicio);
        }
 
        if (0 <= posInicio && posInicio < posFim && posFim < s.length()) {
            resp = s.substring(posInicio, posFim);
        }
 
        return resp;
    }


    public void ler(String nomeArquivo) throws IOException{
        try{
            FileReader file = new FileReader(nomeArquivo);
            BufferedReader buffer = new BufferedReader(file);
            String html = "";
            String line = buffer.readLine();
            while (line != null) {
                html += line;
                line = buffer.readLine();
            }
    
            buffer.close();
            file.close();
    
            html = html.substring(html.indexOf("Full name"));
            html = html.substring(0, html.indexOf("<table style"));
            String campos[] = html.split("<tr>");
    
            this.nomeArquivo = nomeArquivo;

            //Todos os componentes estao em ordem

            for (String campo : campos) {
                // Full name
                if (removeTags(campo).contains("Full name")) {
                    campo = removeTags(campo);
                    this.nome = campo.substring(9).trim();
    
                    // Nickname(s)
                } else if (removeTags(campo).contains("Nickname(s)")) {
                    campo = removeTags(campo);
                    this.apelido = campo.substring(11).trim();
    
                    // Founded
                } else if (removeTags(campo).toLowerCase().contains("founded")) {
                    campo = removeTags(campo.split("<br />")[0]);
                    this.fundacaoMes = this.getMes(campo.toLowerCase());
    
                    if (this.fundacaoMes == 0) {
                        this.fundacaoDia = 0;
                        campo = campo.substring(7);
                        this.fundacaoAno = Integer.parseInt(campo.substring(0, 4));
                    } else {
                        campo = campo.substring(7);
                        String data[] = campo.split(" ");
                        if (data.length < 3) {
                            this.fundacaoDia = 0;
                            this.fundacaoAno = Integer.parseInt(data[1].substring(0, 4));
                        }/* else {
                            if (campo.contains(",")) {
                                this.fundacaoDia = Integer.parseInt(data[1].replace("th", "").replace(",", ""));
                                this.fundacaoAno = Integer.parseInt(data[2].substring(0, 4));
                            } else if (Character.isDigit(data[0].charAt(0))) {
                                this.fundacaoDia = Integer.parseInt(data[0]);
                                this.fundacaoAno = Integer.parseInt(data[2].substring(0, 4));
                            } else {
                                this.fundacaoDia = 0;
                                this.fundacaoAno = Integer.parseInt(data[1].substring(0, 4));
                            }
                        }*/
                    }
    
                    // Ground
                } else if (removeTags(campo).toLowerCase().contains("ground")) {
                    campo = removeTags(campo);
                    this.estadio = estadio.substring(6).trim();
    
                    // Capacity
                } else if (removeTags(campo).toLowerCase().contains("capacity")) {
                    campo = campo.split("<br")[0];
                    campo = removeTags(campo.split("</td>")[0].replace(" ", ""));
                    this.capacidade = Integer.parseInt(removePunctuation(campo.substring(8).split(";")[0]));
    
                    // Coach
                } else if (removeTags(campo).toLowerCase().contains("coach") || campo.toLowerCase().contains("manager")) {
                    campo = removeTags(campo).replace("(es)", "").trim();
                    if (campo.toLowerCase().contains("coach")) {
                        int index = campo.toLowerCase().indexOf("coach");
                        this.tecnico = campo.substring(index + 5).trim();
                    } else if (campo.toLowerCase().contains("manager") && this.tecnico.isEmpty()) {
                        int index = campo.toLowerCase().indexOf("manager");
                        this.tecnico = campo.substring(index + 7).trim();
                    }
                    // League
                } else if (removeTags(campo).contains("League") && this.liga.isEmpty()) {
                    campo = removeTags(campo);
                    this.liga = campo.substring(6).trim();
                }
            }
            File Arq = new File(nomeArquivo);
            this.paginaTam = Arq.length();
            file.close();
        }
        catch(IOException erro){
            throw new IOException("ERRO!");
        }
    }


    public Time clone(){
		Time novo = new Time(nome, apelido, estadio, tecnico, liga, nomeArquivo, capacidade, fundacaoDia, fundacaoMes, fundacaoAno, paginaTam);
		return novo;
    }
    

    public String removeTags(String entrada){
        
        boolean limpo = false;
        int comeco = 0, fim = entrada.length();

        /*while(!limpo){
            if()
        }*/

        return entrada.trim();
    }

    /*
    public static String removerTags(String string, String tag1, String tag2)
	{

        char[] stringChar = string.toCharArray(); 
        int tamanho = stringChar.length;
        boolean isTag1 = false, isTag2 = false;
        
		for(int i = 0; i < string.length(); i++)
		{	 
			if( stringChar[i] == '<' &&
			    stringChar[i+1] == '/' &&
				stringChar[i+2] == 't' &&
				stringChar[i+3] == 'd' &&
				stringChar[i+4] == '>')
			{
				continue;
            }
            
			if(stringChar[i] == tag1.charAt(0))
			{
                isTag1 = true;
                for(int j = 0; j < tag1.length(); j++)
                {
                    if(tag1.charAt(j) != stringChar[i + j])
                    {
                        isTag1 = false;
                    }
                }
                if(isTag1)
                {
                    for(int j = (i + tag1.length() + 1); j < stringChar.length; j++)
                    {
                        if(stringChar[j] == tag2.charAt(0))
                        {
                            isTag2 = true;
                            for(int k = 0; k < tag2.length(); k++)
                            {
                                if(tag2.charAt(k) != stringChar[j + k])
                                {
                                    isTag2 = false;
                                    k = tag2.length();
                                }
                            }
                            if(isTag2)
                            {
                                for(int k = 0; k < j-i +tag2.length(); k++)
                                {
                                    for(int x = i; x < string.length()-1; x++)
                                    {
                                        stringChar[x] = stringChar[x+1];
                                    }

                                    string = new String(stringChar).substring(0, tamanho-1);
                                    stringChar = string.toCharArray();
                                    tamanho--;
                                }
                                
                                i--;
                                j = string.length();
                            }	
                        }
                    }
                }
			}
		}

		return string;
    }
    */
}